Index shard got corrupted

Hi 

in all our elasticsearch cluster we use this elasticsearch-cloud-aws plugin to create the snapshots on s3  on a regular basis.

Some times we saw the shard got corrupted for an index in our elasticsearch log.
So we try to restore it from backup and while restoring it from backup again we see the same exception in logs which is follows

```
[2015-02-25 08:18:10,824][WARN ][indices.cluster          ] [test-es-cluster-1e-data-2] [lst_p113_v_4_20140615_0000][0] failed to start shard
org.elasticsearch.index.gateway.IndexShardGatewayRecoveryException: [lst_p113_v_4_20140615_0000][0] failed recovery
        at org.elasticsearch.index.gateway.IndexShardGatewayService$1.run(IndexShardGatewayService.java:185)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: [lst_p113_v_4_20140615_0000][0] restore failed
        at org.elasticsearch.index.snapshots.IndexShardSnapshotAndRestoreService.restore(IndexShardSnapshotAndRestoreService.java:130)
        at org.elasticsearch.index.gateway.IndexShardGatewayService$1.run(IndexShardGatewayService.java:127)
        ... 3 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: [lst_p113_v_4_20140615_0000][0] failed to restore snapshot [listening-prod6-20150224]
        at org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardRepository.restore(BlobStoreIndexShardRepository.java:165)
        at org.elasticsearch.index.snapshots.IndexShardSnapshotAndRestoreService.restore(IndexShardSnapshotAndRestoreService.java:124)
        ... 4 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: [lst_p113_v_4_20140615_0000][0] Failed to recover index
        at org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardRepository$RestoreContext.restore(BlobStoreIndexShardRepository.java:787)
        at org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardRepository.restore(BlobStoreIndexShardRepository.java:162)
        ... 5 more
Caused by: org.apache.lucene.index.CorruptIndexException: checksum failed (hardware problem?) : expected=1lvsjli actual=3awj8p resource=(org.apache.lucene.store.FSDirectory$FSIndexOutput@7266a49d)
        at org.elasticsearch.index.store.LegacyVerification$Adler32VerifyingIndexOutput.verify(LegacyVerification.java:73)
        at org.elasticsearch.index.store.Store.verify(Store.java:365)
        at org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardRepository$RestoreContext.restoreFile(BlobStoreIndexShardRepository.java:843)
        at org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardRepository$RestoreContext.restore(BlobStoreIndexShardRepository.java:784)
        ... 6 more
[2015-02-25 08:18:10,826][WARN ][cluster.action.shard     ] [test-es-cluster-1e-data-2] [lst_p113_v_4_20140615_0000][0] sending failed shard for [lst_p113_v_4_20140615_0000][0], node[shNgLjr8RlW7Zrk3P4UdPg], [P], restoring[aws-prod-elasticsearch-backup:listening-prod6-20150224], s[INITIALIZING], indexUUID [ZQKQ-6naQqeLP1Gk8IFsig], reason [Failed to start shard, message [IndexShardGatewayRecoveryException[[lst_p113_v_4_20140615_0000][0] failed recovery]; nested: IndexShardRestoreFailedException[[lst_p113_v_4_20140615_0000][0] restore failed]; nested: IndexShardRestoreFailedException[[lst_p113_v_4_20140615_0000][0] failed to restore snapshot [listening-prod6-20150224]]; nested: IndexShardRestoreFailedException[[lst_p113_v_4_20140615_0000][0] Failed to recover index]; nested: CorruptIndexException[checksum failed (hardware problem?) : expected=1lvsjli actual=3awj8p resource=(org.apache.lucene.store.FSDirectory$FSIndexOutput@7266a49d)]; ]] 
```

Even if we back to an older snapshot we found the same exception.

So what we did was we download all the segments files from s3 merge it and there we found some segments were corrupted by using `org.apache.lucene.index.CheckIndex` with `-fix` 
We fixed it but we loose 5gb data.

We shared this problem with aws team, here what they suggested:

> We have taken a look at the ElasticSearch plugin you have identified at https://github.com/elasticsearch/elasticsearch-cloud-aws . In general, as this plugin is not something developed by AWS, our ability to provide support here may be limited. > If you have not already raised this with the developers of this plugin, We would certainly do so in order to follow this up on both sides. 
>
> Now, with regard to the S3 side, there is a feature available in S3 that clients can use to ensure data integrity on upload. Whenever an object is PUT to an S3 bucket, the client is able to supply the 'Content-MD5' header and an appropriate MD5sum matching the contents of the data being uploaded. S3 will verify the data against this MD5 prior to accepting an upload request. Any upload that does not match the supplied MD5 will be rejected. For reference, please see the S3 PutObject API:
>
> http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html
>
> In looking at the ElasticSearch plugin, it is not clear that the client uses this feature. It is certainly not available as a configuration option. From looking at the source code (and I must note, I am not a Java developer), it does not appear this feature is used when objects are uploaded to S3 in:
> 
> https://github.com/elasticsearch/elasticsearch-cloud-aws/blob/master/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java
>
> As a general comment, the scenario you are describing is most commonly encountered when a file is uploaded to S3 while it is open for write by an application. As such, the contents of the file may change while the upload is in progress. This can manifest in corruption, but this is not an S3 problem per-se. Essentially, clients uploading to S3 should ensure the source files are not currently being updated at the time of an upload to ensure data integrity. S3 is an extremely reliable platform, but it can only store the bytes it receives. Supplying the Content-MD5 header, as per my above suggestion, often catches scenarios where a file is being modified as it is in transit, as the uploaded data typically won't match the pre-calculated MD5. 

Can you guys please have a look into this issue and suggest something

Thanks

--- a/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java
+++ b/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java
@@ -21,12 +21,18 @@ package org.elasticsearch.cloud.aws.blobstore;
 
 import com.amazonaws.AmazonClientException;
 import com.amazonaws.services.s3.model.*;
+import com.amazonaws.util.Base64;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
 import org.elasticsearch.common.unit.ByteSizeUnit;
 import org.elasticsearch.common.unit.ByteSizeValue;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
 import java.util.ArrayList;
 import java.util.List;
 
@@ -49,7 +55,7 @@ import java.util.List;
 public class DefaultS3OutputStream extends S3OutputStream {
 
     private static final ByteSizeValue MULTIPART_MAX_SIZE = new ByteSizeValue(5, ByteSizeUnit.GB);
-
+    private static final ESLogger logger = Loggers.getLogger("cloud.aws");
     /**
      * Multipart Upload API data
      */
@@ -120,7 +126,28 @@ public class DefaultS3OutputStream extends S3OutputStream {
             md.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
         }
         md.setContentLength(length);
-        blobStore.client().putObject(bucketName, blobName, is, md);
+
+        InputStream inputStream = is;
+
+        // We try to compute a MD5 while reading it
+        MessageDigest messageDigest;
+        try {
+            messageDigest = MessageDigest.getInstance("MD5");
+            inputStream = new DigestInputStream(is, messageDigest);
+        } catch (NoSuchAlgorithmException impossible) {
+            // Every implementation of the Java platform is required to support MD5 (see MessageDigest)
+            throw new RuntimeException(impossible);
+        }
+        PutObjectResult putObjectResult = blobStore.client().putObject(bucketName, blobName, inputStream, md);
+
+        String localMd5 = Base64.encodeAsString(messageDigest.digest());
+        String remoteMd5 = putObjectResult.getContentMd5();
+        if (!localMd5.equals(remoteMd5)) {
+            logger.debug("MD5 local [{}], remote [{}] are not equal...", localMd5, remoteMd5);
+            throw new AmazonS3Exception("MD5 local [" + localMd5 +
+                    "], remote [" + remoteMd5 +
+                    "] are not equal...");
+        }
     }
 
     private void initializeMultipart() {

